File: src/main/ruby/org/o3project/odenos/remoteobject/message_dispatcher.rb

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #

require 'redis'
require 'thread'
require 'concurrent'
require 'securerandom'
require 'msgpack'
require 'odenos/core/util/logger'

module Odenos
  module Core
    class MessageDispatcher
      TYPE_REQUEST = 0
      TYPE_RESPONSE = 1
      TYPE_EVENT = 2

      include Odenos::Util::Logger
      #
      # Generic Redis class can't change the subscription channel dynamically.
      # add new methods.
      class RedisDynSub < Redis
        def subscribe_add(*channels)
          client.instance_variable_get(:@client).write([['SUBSCRIBE',
                                                         *channels.map(
                                                           &:to_s)]])
        end

        def subscribe_del(*channels)
          client.instance_variable_get(:@client).write([['UNSUBSCRIBE',
                                                         *channels.map(
                                                           &:to_s)]])
        end
      end

      class EventSubscriptionMap
        def initialize
          @_subscriptions = {}
          @_subscription_map = {}
        end

        def get_subscriptions_keys
          @_subscription_map.keys
        end

        def set_subscription(subscription)
          subscriber = subscription.subscriber_id
          subscription.event_filters.each_key do| publisherid |
            subscription.event_filters[publisherid].each do| event_type |
              key = publisherid + ':' + event_type
              unless @_subscription_map.key?(key)
                @_subscription_map[key] = Array.new
              end
              @_subscription_map[key].push(subscriber)
              @_subscription_map[key].uniq!
              unless @_subscriptions.key?(subscriber)
                @_subscriptions[subscriber] = Array.new
              end
              @_subscriptions[subscriber].push(key)
              @_subscriptions[subscriber].uniq!
            end
          end
        end

        def remove_subscription(subscriber)
          if @_subscriptions.key?(subscriber)
            keys = @_subscriptions[subscriber]
          else
            keys = []
          end
          @_subscriptions.delete(subscriber)
          @_subscription_map.each do|key, subscribers|
            subscribers.delete(subscriber)
            if subscribers.size == 0
              @_subscription_map.delete(key)
            end
          end
          keys
        end

        def get_subscribers(publisher_id, event_type)
          key = publisher_id + ':' + event_type
          @_subscription_map[key]
        end

        def clear
          @_subscriptions.clear
          @_subscription_map.clear
        end
      end

      class PublishData
        attr_accessor :trans, :type, :sno, :channel, :data
        def initialize(trans, type, sno, channel, data)
          @channel = channel
          @data = data
          @trans = trans
          @type = type
          @sno = sno
        end
      end

      attr_reader :system_manager_id
      def initialize(system_manager_id = 'systemmanager',
                     redis_server = 'localhost', redis_port = 6379)
        @clients = {}
        @redis_server = redis_server
        @redis_port = redis_port
        @source_dispatcher_id = SecureRandom.uuid
        @system_manager_id = system_manager_id
        @event_manager_id = 'eventmanager'
        @local_objects = {}
        @pubsqueue = Queue.new
        @redis_publisher = nil
        @redis_subscriber_thread = nil
        @subscription_map = EventSubscriptionMap.new
      end
      # start loop
      def start
        @thread_pool = Concurrent::CachedThreadPool.new

        @redis_subscriber_thread ||= Thread.new do
          begin
            @redis_subscriber = RedisDynSub.new(host: @redis_server,
                                                port: @redis_port)
            ch = Array.new
            ch << @source_dispatcher_id
            ch.concat(@local_objects.keys)
            @subscription_map.get_subscriptions_keys.each do |key|
              ch.concat(key)
            end
            @redis_subscriber.subscribe(ch) do |on|
              on.message do |_channel, message|
                @unpacker ||= MessagePack::Unpacker.new
                @unpacker.feed(message)
                type = @unpacker.read
                sno = @unpacker.read
                srcid = @unpacker.read
                case type
                when TYPE_REQUEST
                  request = Request.new(@unpacker.read)
                  @thread_pool << proc do
                    response = dispatch_request(request)
                    debug { "dispatch_request Response:#{response.to_a.inspect}" }
                    packer = MessagePack::Packer.new
                    packer.write(TYPE_RESPONSE)
                    packer.write(sno)
                    packer.write(request.remote_object_id)
                    packer.write(response)
                    packer.flush
                    push_publish_queue(nil, TYPE_RESPONSE,
                                       sno, srcid, packer.to_s)
                  end
                when TYPE_RESPONSE
                  trans = @clients[srcid]
                  response = Response.new(@unpacker.read)
                  trans.signal_response(sno, response)
                when TYPE_EVENT
                  event = Event.new(@unpacker.read)
                  @thread_pool << proc do
                    dispatch_event(event)
                  end
                end
              end
            end
          rescue Redis::BaseConnectionError
            sleep 1
            retry
          ensure
            @redis_subscriber.quit
            @redis_subscriber = nil
          end
        end
        @redis_publisher = Redis.new(host: @redis_server, port: @redis_port)
        @redis_publisher_thread ||= Thread.new do
          begin
            while :true
              pubs = @pubsqueue.pop
              begin
                @redis_publisher.publish(pubs.channel, pubs.data)
              rescue
                if pubs.type == TYPE_REQUEST
                  unless pubs.trans.nil?
                    pubs.trans.signal_response(pubs.sno,
                                               Response.new(Response.NOT_FOUND,
                                                            nil))
                  end
                end
              end
            end
          rescue
            sleep 1
            retry
          end
        end
      end

      def stop
        @thread_pool.shutdown
        if @redis_subscriber_thread
          @redis_subscriber_thread.kill
          @redis_subscriber_thread = nil
        end
        if @redis_publisher_thread
          @redis_publisher.quit
          @redis_publisher_thread.kill
          @redis_publisher_thread = nil
        end
      end

      def close
        stop
        @clients.each_value(&:close)
        @clients.clear
      end

      def set_remote_system_manager
        debug { 'creating RemoteMessageTransport' }
        @clients[@system_manager_id] = RemoteMessageTransport.new(
                                           @system_manager_id,
                                           self)
      end

      # @param [RemoteObject] remote_object
      def add_local_object(remote_object)
        @local_objects[remote_object.remote_object_id] = remote_object
        if @redis_subscriber
          @redis_subscriber.subscribe_add(remote_object.remote_object_id)
        end
      end

      # @param [RemoteObject] remote_object
      def remove_local_object(remote_object)
        @local_objects.delete(remote_object.remote_object_id)
        @subscription_map.remove_subscription(remote_object.remote_object_id)
        if @redis_subscriber
          @redis_subscriber.subscribe_del(remote_object.remote_object_id)
        end
      end

      # @param [String] remote_object_id
      def add_remote_client(remote_object_id)
        @clients[remote_object_id] = RemoteMessageTransport.new(
                                         remote_object_id,
                                         self)
      end

      # @param [String] remote_object_id
      def remove_remote_client(remote_object_id)
        @clients.delete(remote_object_id)
      end

      # @param [Request] request
      def request_sync(request)
        client = get_message_client(request.remote_object_id)
        client.send_request_message(request)
      end

      # @param [Event] event
      def publish_event_async(event)
        packer = MessagePack::Packer.new
        packer.write(TYPE_EVENT)
        packer.write(0)
        packer.write(@source_dispatcher_id)
        packer.write(event)
        packer.flush
        push_publish_queue(nil, TYPE_EVENT, 0,
                           event.publisher_id + ':' + event.event_type,
                           packer.to_s)
      end
      # @param [EventSubscription] event_subscription
      # @return [Future]
      def subscribe_event(event_subscription)
        debug { "subscribe_event( #{event_subscription.to_hash.inspect} )" }
        @subscription_map.set_subscription(event_subscription)
        @subscription_map.get_subscriptions_keys.each do|key|
          @redis_subscriber.subscribe_add(key) if @redis_subscriber
        end

        request = Request.new(event_manager_id, :PUT,
                              "settings/event_subscriptions/#{event_manager_id}",
                              event_subscription)
        request_sync(request)

        Response.new(Response::OK, nil)
      end

      def push_publish_queue(trans, type, sno, chanel, data)
        @pubsqueue.push(PublishData.new(trans, type, sno, chanel, data))
      end
      # @param [Request] request
      def dispatch_request(request)
        debug { "dispatch_request( #{request.to_a.inspect} )" }

        if @local_objects.key?(request.remote_object_id)
          object = @local_objects[request.remote_object_id]
          return object.dispatch_request(request)
        end
        Response.new(404, nil)
      end
      # @param [Event] event
      def dispatch_event(event)
        debug { "dispatch_event( #{event.to_a.inspect} )" }
        del_obj_ids = []
        subscribers = @subscription_map.get_subscribers(event.publisher_id,
                                                        event.event_type)
        subscribers.each do|subscriber|
          if @local_objects.key?(subscriber)
            object = @local_objects[subscriber]
            if object.nil?
              del_obj_ids.push(subscriber)
            else
              object.dispatch_event(event)
            end
          end
          for id in del_obj_ids do
            @local_objects.delete(id)
            @subscription_map.remove_subscription(id)
          end
        end
      end
      # @!attribute [r] event_manager_id
      #   @return [String] Event Manager's RemoteObject ID
      attr_reader :event_manager_id

      def get_redis_publisher
        @redis_publisher
      end

      def get_source_dispatcher_id
        @source_dispatcher_id
      end

      protected

      # @return [BaseMessageTransport]
      def get_message_client(remote_object_id)
        unless @clients.include? remote_object_id
          if @local_objects.include? remote_object_id
            add_local_object LocalMessageTransport.new(remote_object_id, self)
          else
            add_remote_client(remote_object_id)
          end
        end
        @clients[remote_object_id]
      end
    end
  end
end

File: src/main/python/org/o3project/odenos/remoteobject/transport/message_dispatcher.py
# -*- coding:utf-8 -*-

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #

import redis
import threading
import uuid
from io import BytesIO
import Queue
import msgpack
import logging
import traceback
from concurrent import futures

from org.o3project.odenos.remoteobject.message.response import Response
from org.o3project.odenos.remoteobject.message.request import Request
from org.o3project.odenos.remoteobject.message.event import Event

from org.o3project.odenos.remoteobject.transport.local_message_transport import (
    LocalMessageTransport
)
from org.o3project.odenos.remoteobject.transport.remote_message_transport import (
    RemoteMessageTransport
)


class MessageDispatcher:
    TYPE_REQUEST = 0
    TYPE_RESPONSE = 1
    TYPE_EVENT = 2
    REDIS_SERVER = "localhost"
    REDIS_PORT = 6379
    SYSTEM_MANAGER_ID = "systemmanager"
    TIME_RETRY_INTERVAL = 100

    class EventSubscriptionMap:

        def __init__(self):
            self.__subscriptions = {}
            self.__subscription_map = {}

        def get_subscriptions_keys(self):
            return self.__subscription_map.keys()

        def set_subscription(self, subscription):
            subscriber = subscription.subscriber_id
            for publisherid in subscription.event_filters.keys():
                for event_type in subscription.event_filters[publisherid]:
                    key = publisherid + ":" + event_type
                    if key not in self.__subscription_map:
                        self.__subscription_map[key] = set()
                    self.__subscription_map[key].add(subscriber)
                    if subscriber not in self.__subscriptions:
                        self.__subscriptions[subscriber] = set()
                    self.__subscriptions[subscriber].add(key)

        def remove_subscription(self, subscriber):
            if subscriber in self.__subscriptions:
                keys = self.__subscriptions[subscriber]
                del self.__subscriptions[subscriber]
            else:
                keys = set()
            for key in self.__subscription_map.copy():
                subscribers = self.__subscription_map[key]
                if subscriber in subscribers:
                    subscribers.remove(subscriber)
                if len(subscribers) == 0:
                    del self.__subscription_map[key]
            return keys

        def get_subscribers(self, *args):
            if len(args) == 0:
                return set(self.__subscriptions.keys())
            else:
                publisher_id, event_type = args
                key = publisher_id + ":" + event_type
                subscribers = self.__subscription_map[key]
                return subscribers

        def clear(self):
            self.__subscriptions.clear()
            self.__subscription_map.clear()

    class PublishData:

        def __init__(self, trans, type_, sno, channel, data):
            self.trans = trans
            self.type = type_
            self.sno = sno
            self.channel = channel
            self.data = data

    def __init__(self, system_manager_id=SYSTEM_MANAGER_ID,
                 redis_server=REDIS_SERVER,
                 redis_port=REDIS_PORT):
        self.__clients = {}
        self.__local_objects = {}
        self.__event_manager_id = None
        self.system_manager_id = system_manager_id
        self.__redisServer = redis_server
        self.__redisPort = redis_port
        self.__sourceDispatcherId = str(uuid.uuid4())
        self.__pubsqueue = Queue.Queue()
        self.__redisSubscriberThread = None
        self.__redisSubscriber = None
        self.__redisPublisherThread = None
        self.__redisPubliser = None
        self.__subscription_map = self.EventSubscriptionMap()

    def __redisPublisherRunnable(self):
        while (True):
            pubs = self.__pubsqueue.get()
            if pubs is None:
                break
            try:
                self.__redisPubliser.publish(pubs.channel, pubs.data)
            except:
                if pubs.type == self.TYPE_REQUEST:
                    if pubs.trans is not None:
                        pubs.trans.signalResponse(
                            pubs.sno,
                            Response(Response.StatusCode.NOT_FOUND, None))
                pass

    def __redisSubscriberRunnable(self):
        for mesg in self.__redisSubscriber.listen():
            if mesg['type'] != 'message':
                continue
            try:
                bio = BytesIO()
                bio.write(mesg['data'])
                bio.seek(0)
                upk = msgpack.Unpacker(bio)
                tp = upk.unpack()
                sno = upk.unpack()
                srcid = upk.unpack()
                if tp == self.TYPE_REQUEST:
                    request = Request.create_from_packed(upk.unpack())

                    def request_runnable(request, sno, srcid):
                        try:
                            response = self.dispatch_request(request)
                            pk = msgpack.Packer()
                            resb = bytearray()
                            resb.extend(pk.pack(self.TYPE_RESPONSE))
                            resb.extend(pk.pack(sno))
                            resb.extend(pk.pack(request.object_id))
                            resb.extend(pk.pack(response.packed_object()))
                            self.__pubsqueue.put(
                                MessageDispatcher.PublishData(None,
                                                              self.TYPE_RESPONSE,
                                                              sno,
                                                              srcid,
                                                              resb))
                        except:
                            logging.exception('Request processing error')
                    self.thread_pool.submit(request_runnable,
                                            request,
                                            sno,
                                            srcid)
                elif tp == self.TYPE_RESPONSE:
                    trans = self.__clients[srcid]
                    response = Response.create_from_packed(upk.unpack())
                    trans.signalResponse(sno, response)
                elif tp == self.TYPE_EVENT:
                    event = Event.create_from_packed(upk.unpack())

                    def event_runnable(event):
                        try:
                            self.dispatch_event(event)
                        except:
                            logging.exception('Event processing error')
                    self.thread_pool.submit(event_runnable, event)
            except:
                logging.error(traceback.format_exc())
                pass

    def start(self):
        self.thread_pool = futures.ThreadPoolExecutor(max_workers=8)
        self.__redisPubliser = redis.StrictRedis(host=self.__redisServer,
                                                 port=self.__redisPort)
        self.__redisSubscriber = redis.StrictRedis(
            host=self.__redisServer,
            port=self.__redisPort).pubsub()
        self.__update_subscriber()

        self.__redisSubscriberThread = threading.Thread(
            target=self.__redisSubscriberRunnable)
        self.__redisSubscriberThread.start()

        self.__redisPublisherThread = threading.Thread(
            target=self.__redisPublisherRunnable)
        self.__redisPublisherThread.start()

    def stop(self):
        if not (self.__redisSubscriberThread is None):
            self.__redisSubscriber.unsubscribe()
            self.__redisSubscriberThread.join()
            self.__redisSubscriberThread = None
        if not (self.__redisPublisherThread is None):
            self.__pubsqueue.put(None)
            self.__redisPublisherThread.join()
            self.__redisPublisherThread = None

    def join(self):
        if not (self.__redisSubscriberThread is None):
            self.__redisSubscriberThread.join()
            self.__redisSubscriberThread = None
        if not (self.__redisPublisherThread is None):
            self.__redisPublisherThread.join()
            self.__redisPublisherThread = None

    def close(self):
        self.stop()
        self.__local_objects.clear()
        self.__clients.clear()

    def __get_message_client(self, object_id):
        if object_id not in self.__clients:
            if object_id in self.__local_objects:
                client = LocalMessageTransport(object_id, self)
            else:
                client = RemoteMessageTransport(object_id, self)
            self.__clients[object_id] = client
        return self.__clients[object_id]

    def get_source_dispatcher_id(self):
        return self.__sourceDispatcherId

    def getRedisPublisher(self):
        return self.__redisPubliser

    def __update_subscriber(self):
        if self.__redisSubscriber is not None:
            keys = []
            keys.append(self.__sourceDispatcherId)
            for id_ in self.__local_objects.keys():
                keys.append(id_)
            for key in self.__subscription_map.get_subscriptions_keys():
                keys.append(key)
            self.__redisSubscriber.subscribe(keys)

    @property
    def event_manager_id(self):
        if not self.__event_manager_id:
            # TODO resolve id from SystemManager
            self.__event_manager_id = "eventmanager"
        return self.__event_manager_id

    def add_local_object(self, remote_object):
        self.__local_objects[remote_object.object_id] = remote_object
        self.__update_subscriber()

    def remove_local_object(self, remote_object):
        del self.__local_objects[remote_object.object_id]
        self.__subscription_map.remove_subscription(
            remote_object.object_id)
        self.__update_subscriber()

    def add_remote_client(self, object_id):
        client = RemoteMessageTransport(object_id, self)
        self.__clients[object_id] = client

    def remove_remote_client(self, object_id):
        self.__clients.pop(object_id)

    def get_local_object_ids(self):
        return self.__local_objects.keys()

    def set_remote_system_manager(self):
        self.__clients[self.system_manager_id] = \
            RemoteMessageTransport(self.system_manager_id, self)

    def request_sync(self, request):
        client = self.__get_message_client(request.object_id)
        logging.debug("[request_sync ]:send to " + client.object_id)
        return client.send_request_message(request)

    def publish_event_async(self, event):
        pk = msgpack.Packer()
        resb = bytearray()
        resb.extend(pk.pack(self.TYPE_EVENT))
        resb.extend(pk.pack(0))
        resb.extend(pk.pack(event.publisher_id + ":" + event.event_type))
        resb.extend(pk.pack(event.packed_object()))
        self.pushPublishQueue(None,
                              self.TYPE_EVENT,
                              0,
                              event.publisher_id + ":" + event.event_type,
                              resb)

    def subscribe_event(self, event_subscription):
        self.__subscription_map.remove_subscription(
            event_subscription.subscriber_id)
        self.__subscription_map.set_subscription(event_subscription)

        self.__update_subscriber()

        request = Request(self.event_manager_id,
                          Request.Method.PUT,
                          "settings/event_subscriptions/%s" % event_subscription.subscriber_id,
                          event_subscription)
        return self.request_sync(request)

    def pushPublishQueue(self, trans, type_, sno, channel, data):
        self.__pubsqueue.put(MessageDispatcher.PublishData(trans,
                                                           type_,
                                                           sno,
                                                           channel,
                                                           data))

    def dispatch_request(self, request):
        if request.object_id in self.__local_objects:
            return self.__local_objects[request.object_id].dispatch_request(
                request)
        else:
            return Response(404, None)

    def dispatch_event(self, event):
        subscribers = self.__subscription_map.get_subscribers(
            event.publisher_id,
            event.event_type)
        del_obj_ids = []
        for es in subscribers.copy():
            if es in self.__local_objects:
                if self.__local_objects[es] is None:
                    del_obj_ids.append(es)
                    continue
                self.__local_objects[es].dispatch_event(event)
        for id in del_obj_ids:
            del self.__local_objects[id]
            self.__subscription_map.remove_subscription(id)
